overlord/snapshotstate/snapshotstate.go

```diff
@@ -0,0 +1,264 @@
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+type snapshotState struct {
+	LastID uint64 `json:"last-id"`
+}
+
+func newSnapshotID(st *state.State) (uint64, error) {
+	var shotState snapshotState
+
+	err := st.Get("snapshot", &shotState)
+	if err != nil {
+		if err == state.ErrNoState {
+			shotState = snapshotState{}
+		} else {
+			return 0, err
+		}
+	}
+
+	shotState.LastID++
+	st.Set("snapshot", shotState)
+
+	return shotState.LastID, nil
+}
+
+func allActiveSnapNames(st *state.State) ([]string, error) {
+	all, err := snapstate.All(st)
+	if err != nil {
+		return nil, err
+	}
+	names := make([]string, 0, len(all))
+	for name, snapst := range all {
+		if snapst.IsInstalled() {
+			names = append(names, name)
+		}
+	}
+
+	return names, nil
+}
+
+func snapNamesInSnapshot(snapshotID uint64, requested []string) (snapsFound []string, filenames []string, err error) {
+	sort.Strings(requested)
+	found := false
+	err = backend.Iter(context.Background(), func(r *backend.Reader) error {
+		if r.SetID == snapshotID {
+			found = true
+			if len(requested) == 0 || strutil.SortedListContains(requested, r.Snap) {
+				snapsFound = append(snapsFound, r.Snap)
+				filenames = append(filenames, r.Filename())
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	if !found {
+		return nil, nil, client.ErrSnapshotNotFound
+	}
+	if len(snapsFound) == 0 {
+		return nil, nil, client.ErrSnapshotSnapsNotFound
+	}
+
+	return snapsFound, filenames, nil
+}
+
+// some snapshot ops only conflict with other snapshot ops (in
+// particular check, forget and restore)
+func checkSnapshotChangeConflict(st *state.State, snapshotID uint64, conflictingKinds ...string) error {
+	for _, task := range st.Tasks() {
+		chg := task.Change()
+		if chg.Status().Ready() {
+			continue
+		}
+		if !strutil.ListContains(conflictingKinds, chg.Kind()) {
+			continue
+		}
+
+		var shot shotState
+		if err := task.Get("shot", &shot); err != nil {
+			return err
+		}
+
+		if shot.ID == snapshotID {
+			return fmt.Errorf("snapshot #%d has a %q change in progress", snapshotID, chg.Kind())
+		}
+	}
+
+	return nil
+}
+
+// List valid snapshots
+//
+// The snapshots are closed before returning.
+var List = backend.List
+
+// Save creates a taskset for taking snapshots of snaps' data.
+// Note that the state must be locked by the caller.
+func Save(st *state.State, snapNames []string, users []string) (uint64, []string, *state.TaskSet, error) {
+	var err error
+
+	if len(snapNames) == 0 {
+		snapNames, err = allActiveSnapNames(st)
+		if err != nil {
+			return 0, nil, nil, err
+		}
+	}
+
+	if err := snapstate.CheckChangeConflictMany(st, snapNames, nil); err != nil {
+		return 0, nil, nil, err
+	}
+
+	shID, err := newSnapshotID(st)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+
+	ts := state.NewTaskSet()
+
+	for _, name := range snapNames {
+		desc := fmt.Sprintf("Saving the data of snap %q in snapshot #%d", name, shID)
```

Comment:
"Save data of snap %q in snapshot set #%d".. task summaries are generally imperative, and ideally a bit shorter.